File: tests/rosetta/transpiler/FS/chernicks-carmichael-numbers.fs
// Generated 2025-07-27 23:36 +0700

exception Break
exception Continue

exception Return

let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let rec isPrime (n: int) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable n = n
    try
        if n < 2 then
            __ret <- false
            raise Return
        if (((n % 2 + 2) % 2)) = 0 then
            __ret <- n = 2
            raise Return
        if (((n % 3 + 3) % 3)) = 0 then
            __ret <- n = 3
            raise Return
        let mutable d: int = 5
        while (d * d) <= n do
            if (((n % d + d) % d)) = 0 then
                __ret <- false
                raise Return
            d <- d + 2
            if (((n % d + d) % d)) = 0 then
                __ret <- false
                raise Return
            d <- d + 4
        __ret <- true
        raise Return
        __ret
    with
        | Return -> __ret
let rec bigTrim (a: int array) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable a = a
    try
        let mutable n: int = Array.length a
        while (n > 1) && ((int (a.[n - 1])) = 0) do
            a <- unbox<int array> (Array.sub a 0 ((n - 1) - 0))
            n <- n - 1
        __ret <- unbox<int array> a
        raise Return
        __ret
    with
        | Return -> __ret
let rec bigFromInt (x: int) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable x = x
    try
        if x = 0 then
            __ret <- unbox<int array> [|0|]
            raise Return
        let mutable digits: int array = [||]
        let mutable n: int = x
        while n > 0 do
            digits <- unbox<int array> (Array.append digits [|((n % 10 + 10) % 10)|])
            n <- n / 10
        __ret <- unbox<int array> digits
        raise Return
        __ret
    with
        | Return -> __ret
let rec bigMulSmall (a: int array) (m: int) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable a = a
    let mutable m = m
    try
        if m = 0 then
            __ret <- unbox<int array> [|0|]
            raise Return
        let mutable res: int array = [||]
        let mutable carry: int = 0
        let mutable i: int = 0
        while i < (int (Array.length a)) do
            let mutable prod = (int ((int (a.[i])) * m)) + carry
            res <- unbox<int array> (Array.append res [|((prod % 10 + 10) % 10)|])
            carry <- int ((int prod) / 10)
            i <- i + 1
        while carry > 0 do
            res <- unbox<int array> (Array.append res [|((carry % 10 + 10) % 10)|])
            carry <- carry / 10
        __ret <- bigTrim res
        raise Return
        __ret
    with
        | Return -> __ret
let rec bigToString (a: int array) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable a = a
    try
        let mutable s: string = ""
        let mutable i: int = (int (Array.length a)) - 1
        while i >= 0 do
            s <- s + (string (a.[i]))
            i <- i - 1
        __ret <- s
        raise Return
        __ret
    with
        | Return -> __ret
let rec pow2 (k: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable k = k
    try
        let mutable r: int = 1
        let mutable i: int = 0
        while i < k do
            r <- r * 2
            i <- i + 1
        __ret <- r
        raise Return
        __ret
    with
        | Return -> __ret
let rec ccFactors (n: int) (m: int) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable n = n
    let mutable m = m
    try
        let mutable p: int = (6 * m) + 1
        if not (isPrime p) then
            __ret <- Array.empty<int>
            raise Return
        let mutable prod: int array = bigFromInt p
        p <- (12 * m) + 1
        if not (isPrime p) then
            __ret <- Array.empty<int>
            raise Return
        prod <- bigMulSmall prod p
        let mutable i: int = 1
        while i <= (n - 2) do
            p <- (int ((int ((int (pow2 i)) * 9)) * m)) + 1
            if not (isPrime p) then
                __ret <- Array.empty<int>
                raise Return
            prod <- bigMulSmall prod p
            i <- i + 1
        __ret <- unbox<int array> prod
        raise Return
        __ret
    with
        | Return -> __ret
let rec ccNumbers (start: int) (``end``: int) =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    let mutable start = start
    let mutable ``end`` = ``end``
    try
        let mutable n: int = start
        try
            while n <= ``end`` do
                try
                    let mutable m: int = 1
                    if n > 4 then
                        m <- pow2 (n - 4)
                    try
                        while true do
                            try
                                let num: int array = ccFactors n m
                                if (int (Array.length num)) > 0 then
                                    printfn "%s" ((("a(" + (string n)) + ") = ") + (unbox<string> (bigToString num)))
                                    raise Break
                                if n <= 4 then
                                    m <- m + 1
                                else
                                    m <- m + (int (pow2 (n - 4)))
                            with
                            | Continue -> ()
                            | Break -> raise Break
                    with
                    | Break -> ()
                    | Continue -> ()
                    n <- n + 1
                with
                | Continue -> ()
                | Break -> raise Break
        with
        | Break -> ()
        | Continue -> ()
        __ret
    with
        | Return -> __ret
ccNumbers 3 9
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

File: tests/rosetta/transpiler/php/chernicks-carmichael-numbers.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _intdiv($a, $b) {
    if (function_exists('bcdiv')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return intval(bcdiv($sa, $sb, 0));
    }
    return intdiv($a, $b);
}
$__start_mem = memory_get_usage();
$__start = _now();
  function isPrime($n) {
  if ($n < 2) {
  return false;
}
  if ($n % 2 == 0) {
  return $n == 2;
}
  if ($n % 3 == 0) {
  return $n == 3;
}
  $d = 5;
  while ($d * $d <= $n) {
  if ($n % $d == 0) {
  return false;
}
  $d = $d + 2;
  if ($n % $d == 0) {
  return false;
}
  $d = $d + 4;
};
  return true;
};
  function bigTrim($a) {
  $n = count($a);
  while ($n > 1 && $a[$n - 1] == 0) {
  $a = array_slice($a, 0, $n - 1 - 0);
  $n = $n - 1;
};
  return $a;
};
  function bigFromInt($x) {
  if ($x == 0) {
  return [0];
}
  $digits = [];
  $n = $x;
  while ($n > 0) {
  $digits = array_merge($digits, [$n % 10]);
  $n = _intdiv($n, 10);
};
  return $digits;
};
  function bigMulSmall($a, $m) {
  if ($m == 0) {
  return [0];
}
  $res = [];
  $carry = 0;
  $i = 0;
  while ($i < count($a)) {
  $prod = $a[$i] * $m + $carry;
  $res = array_merge($res, [$prod % 10]);
  $carry = _intdiv($prod, 10);
  $i = $i + 1;
};
  while ($carry > 0) {
  $res = array_merge($res, [$carry % 10]);
  $carry = _intdiv($carry, 10);
};
  return bigTrim($res);
};
  function bigToString($a) {
  $s = '';
  $i = count($a) - 1;
  while ($i >= 0) {
  $s = $s . _str($a[$i]);
  $i = $i - 1;
};
  return $s;
};
  function pow2($k) {
  $r = 1;
  $i = 0;
  while ($i < $k) {
  $r = $r * 2;
  $i = $i + 1;
};
  return $r;
};
  function ccFactors($n, $m) {
  $p = 6 * $m + 1;
  if (!isPrime($p)) {
  return [];
}
  $prod = bigFromInt($p);
  $p = 12 * $m + 1;
  if (!isPrime($p)) {
  return [];
}
  $prod = bigMulSmall($prod, $p);
  $i = 1;
  while ($i <= $n - 2) {
  $p = (pow2($i) * 9 * $m) + 1;
  if (!isPrime($p)) {
  return [];
}
  $prod = bigMulSmall($prod, $p);
  $i = $i + 1;
};
  return $prod;
};
  function ccNumbers($start, $end) {
  $n = $start;
  while ($n <= $end) {
  $m = 1;
  if ($n > 4) {
  $m = pow2($n - 4);
}
  while (true) {
  $num = ccFactors($n, $m);
  if (count($num) > 0) {
  echo rtrim('a(' . _str($n) . ') = ' . bigToString($num)), PHP_EOL;
  break;
}
  if ($n <= 4) {
  $m = $m + 1;
} else {
  $m = $m + pow2($n - 4);
}
};
  $n = $n + 1;
};
};
  ccNumbers(3, 9);
$__end = _now();
$__end_mem = memory_get_usage();
$__duration = intdiv($__end - $__start, 1000);
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;;
